vmm/src/acpi.rs

```diff
@@ -125,14 +137,64 @@ pub fn create_acpi_tables(
         .expect("Error writing MCFG table");
     tables.push(mcfg_offset.0);
 
+    // SRAT
+    // Only created if the NUMA nodes list is not empty.
+    let numa_nodes = memory_manager.lock().unwrap().numa_nodes().clone();
+    let (prev_tbl_len, prev_tbl_off) = if numa_nodes.is_empty() {
+        (mcfg.len(), mcfg_offset)
+    } else {
+        let mut srat = SDT::new(*b"SRAT", 36, 3, *b"CLOUDH", *b"CHSRAT  ", 1);
+        // SRAT reserved 12 bytes
+        srat.append_slice(&[0u8; 12]);
+
+        for (node_id, node) in numa_nodes.iter() {
+            for region in node.memory_regions() {
+                let proximity_domain = *node_id as u32;
+                let base_addr = region.start_addr().raw_value();
+                let base_addr_lo = (base_addr & 0xffff_ffff) as u32;
+                let base_addr_hi = (base_addr >> 32) as u32;
+                let length = region.len() as u64;
+                let length_lo = (length & 0xffff_ffff) as u32;
+                let length_hi = (length >> 32) as u32;
+
+                // Flags
+                // - Enabled = 1 (bit 0)
+                // - Hot Pluggable = 0 (bit 1)
+                // - NonVolatile = 0 (bit 2)
+                // - Reserved bits 3-31
+                let flags = 1;
+
+                srat.append(MemoryAffinity {
+                    type_: 1,
+                    length: 40,
```

Comment:
Maybe it is worth mentioning those two values are hard-coded as instructed by the ACPI spec? I was assuming the `length` is the `std::mem::size_of::<MemoryAffinity>()`...

Reply:
Better to follow the spec here, instead of getting the size from `MemoryAffinity` as this could hide a bug at some point if we were to change the `MemoryAffinity` structure.
I can add a comment but everything in this specific file assumes you need an ACPI spec next to you to understand how these tables are built :)

Reply:
Right. I found my assumption above was wrong when I saw the ACPI spec afterwards. That's why I thought it might worth adding a comments here to emphasize everything is from the spec.

Reply:
Add `assert!(std::mem::sizeof<MemoryAffinity>::(), 40);`

Reply:
Sounds good